Add package scanner into moderation process

gep13 opened the pull request:
There are still some things that need to be done before this PR can be merged in, but it is ready for a review.

I have successfully walked a package through the happy path, i.e. submit package, pass validator, pass verifier, pass scanner, and then have the package approved by a moderator.  I have also tested this when a package is trusted.  Things that still need to be done.

- [ ] When a error occurs on the package page, i.e. when attempting to approve a package, the validation summary text is in small black text, and hard to see.  This needs to be changed.
- [ ] We should use a modal popup when something goes wrong, to indicate there is a problem.  Similar to what is shown when it updates successsfully
- [ ] We need to decide on exactly what causes a package to be pushed back to the maintainer, and when it goes to human moderation.  At the minute, a package is never pushed back to a maintainer based on scanner results
- [ ] Add input textbox on package page for validation exemption message
- [ ] Add input textbox on package page for scanner exemption message
- [ ] Add link in status summary section to the virus scan results.  Similar to the `Details` hyperlink shown for verification testing URL
- [ ] Show any files that were not scanned by the package-scanner (**NOTE:** This may have to be broken out into a separate scope of work
- [ ] Need to update disclaimer about virus scanning only happening on en-US locale

Fixes #877

gep13 commented:
Here is the happy path...

Package submitted to chocolatey.org...

![image](https://user-images.githubusercontent.com/1271146/87806128-4f337c80-c84e-11ea-8557-1b83d23976cb.png)

Package passes validator...

![image](https://user-images.githubusercontent.com/1271146/87806193-65413d00-c84e-11ea-8e20-06f68dc147e0.png)

Package passes verifier...

![image](https://user-images.githubusercontent.com/1271146/87806260-7c802a80-c84e-11ea-991c-085180ea2a1e.png)

Package passes scanner and is approved (this is a trusted package)...

![image](https://user-images.githubusercontent.com/1271146/87806323-99b4f900-c84e-11ea-9316-006bcedc154e.png)

gep13 commented:
Validation Summary styling has been re-applied.  Now looks like this:

![image](https://user-images.githubusercontent.com/1271146/88685704-20d75c00-d0ee-11ea-9650-b50f7554a665.png)

When there is a problem

gep13 commented:
An error modal will now be shown when something went wrong with the requested operation:

![image](https://user-images.githubusercontent.com/1271146/88686997-9d1e6f00-d0ef-11ea-9adc-e8eae6bec1b3.png)

vexx32 commented on chocolatey/Website/Views/Packages/DisplayPackage.cshtml:
```diff
@@ -559,6 +562,50 @@
+                            <div>
+                                <p class="mb-0">Scan Testing Resulted in @Model.PackageScanResultStatus</p>
+
+                                @if (Model.PackageScanFlagResultType == PackageScanFlagResultType.Note)
+                                {
+                                    <p>At least one file within this package had greater then 0 detections for a file</p>
+                                }
+                                else if (Model.PackageScanFlagResultType == PackageScanFlagResultType.Warning)
+                                {
+                                    <p>At least one file within this package had between 5 and 10 detections for a file</p>
+                                }
+                                else if (Model.PackageScanFlagResultType == PackageScanFlagResultType.Error)
+                                {
+                                    <p>At least one file within this package had greater then 10 detections for a file</p>
+                                }
```
Can this be written as a `switch` statement? I get the impression it would be clearer than this stack of if/elseif

gep13 replied:
:+1: This has been changed.

vexx32 commented on chocolatey/Website/Views/Packages/DisplayPackage.cshtml:
```diff
@@ -559,6 +562,50 @@
+                                }
+                            </div>
+                        }
+                        @if (Model.PackageScanResultStatus == PackageScanStatusType.NotFlagged)
+                        {
+                            <p class="mb-0"><span class="fas fa-check text-success mr-3"></span> Scan Testing Successful - No detections found in any package files</p>
+                        }
+                        @if (Model.PackageScanResultStatus == PackageScanStatusType.Pending)
+                        {
+                            <p class="mb-0"><span class="fas fa-clock text-warning mr-3"></span> Scan Testing Pending</p>
+                        }
+                        @if (Model.PackageScanResultStatus == PackageScanStatusType.Unknown)
+                        {
+                            <p class="mb-0"><span class="fas fa-question text-secondary mr-3"></span> Scan Testing Unknown</p>
+                        }
+                        @if (Model.PackageScanResultStatus == PackageScanStatusType.Exempted)
+                        {
+                            <span class="fas fa-ban text-secondary mr-3"></span>
+                            <div>
+                                <p class="mb-0">Scan Testing Exempt:</p>
+                                @Html.Raw(Markdown.ToHtml(Model.ExemptedFromScannerReason.clean_html() ?? string.Empty, MarkdownPipeline))
+                            </div>
+                        }
```
These statuses are mutually exclusive, right? Should these be `else if`s instead?

Might also be a good candidate for a switch statement.

gep13 replied:
Perhaps it can, but I am a big fan of consistency in the code base, and currently, this is how the other two sections of the code base are currently implemented: https://github.com/chocolatey/chocolatey.org/blob/master/chocolatey/Website/Views/Packages/DisplayPackage.cshtml#L511-L560

vexx32 replied:
I agree that consistency is valuable, but IMO it's better to not waste the compute cycles needlessly re-checking things repeatedly that have already satisfied the only case they're going to satisfy.

Up to you, we can always do a pass over all the cases like this in a follow up PR or something.

gep13 replied:
I have created a follow up issue here: https://github.com/chocolatey/chocolatey.org/issues/942 to capture this suggestion.

AdmiringWorm commented on chocolatey/Website/Controllers/ApiController.cs:
```diff
@@ -553,6 +553,21 @@ public virtual ActionResult SubmitPackageScanResults(string apiKey, string id, s
             }
 
+            PackageScanFlagResultType packageScanFlagResult;
+            try
+            {
+                Enum.TryParse(scanFlagResult.to_string(), true, out packageScanFlagResult);
+            }
+            catch (Exception)
+            {
+                packageScanFlagResult = PackageScanFlagResultType.Unknown;
+            }
```
Only if the out parameter isn't of an enum type.
Would probably be better to replace the try/catch statements with for instance.
```suggestion
            if (!Enum.TryParse(scanFlagResult.to_string(), true, out packageScanFlagResult))
            {
                packageScanFlagResult = PackageScanFlagResultType.Unknown;
            }
```

ferventcoder replied:
> Would probably be better to replace the try/catch statements with for instance.

If it comes back with true/false (like one would expect it to as it is a "Try" method), then this adjustment would be appropriate. However if you give it null/empty/not a string representation of the enum, and it can throw an error, then your suggested change is incorrect.

AdmiringWorm replied:
> However if you give it null/empty/not a string representation of the enum, and it can throw an error, then your suggested change is incorrect.

I agree, if that had been the case trying to parse without a try/catch block would be incorrect. Thankfully, that isn't the case though, and since the `to_string` extension method also does a null check of the object, there shouldn't be a problem there either.

![enum-parsing](https://user-images.githubusercontent.com/1474648/88978402-ee7d5880-d2bf-11ea-971b-ba527e8762fa.gif)

gep13 replied:
This addition was based on the code that is already in place for parsing exisitng enums, so I adapted the same logic for this new enum: https://github.com/chocolatey/chocolatey.org/blob/master/chocolatey/Website/Controllers/ApiController.cs#L607-L615

vexx32 replied:
Perhaps best left as a refactoring issue to be dealt with separately, then? Either way we should probably cut the unneeded try/catches, no sense maintaining unreachable code paths. 🙂

ferventcoder replied:
Are you compiling that to .NET 4.x? Just curious.

AdmiringWorm replied:
Yes. I compiled and tested it while targeting .NET 4.0 Client Profile, 4.5, and 4.8 (not at the same time obviously 😕).

vexx32 replied:
Done! #941

ferventcoder commented:
> Validation Summary styling has been re-applied. Now looks like this:
> 
> ![image](https://user-images.githubusercontent.com/1271146/88685704-20d75c00-d0ee-11ea-9650-b50f7554a665.png)
> 
> When there is a problem

This is a different issue IMHO

gep13 commented:
>@ferventcoder said...
This is a different issue IMHO - can it get its own issue number?

I have created associated issues #916 and #917

gep13 commented on chocolatey/Website/Content/packages.css:
```diff
@@ -136,4 +136,28 @@ textarea {
+    border-left: 6px solid #f4516c;
+}
+
+.validation-summary-errors ul li:not(:first-child) {
+    margin-top: .5rem;
+}
```
I will need to defer to @designxsteph on this one.  I unashamedly stole the css that I have put in here from this file: https://github.com/chocolatey/chocolatey.org/blob/master/chocolatey/Website/Content/account.css#L32-L50

I suspect that during a refactoring it was removed, so I have placed it into the package.css file, which is already included on the page.

vexx32 replied:
Fair enough! I tend to be a little wary of using `:not()` clauses and such, especially in combination with other pseudo-classes. Might just be a pet peeve, but I have a vague recollection that they're slower for browsers to process/render (though I'm sure not by a great deal, so... eh).

gep13 commented on chocolatey/Website/Controllers/PackagesController.cs:
```diff
@@ -145,20 +148,42 @@ public virtual ActionResult DisplayPackage(string id, string version, FormCollec
+                (package.PackageValidationResultStatus == PackageAutomatedReviewResultStatusType.Passing || package.PackageValidationResultStatus == PackageAutomatedReviewResultStatusType.Exempted)
+                    && (package.PackageTestResultStatus == PackageAutomatedReviewResultStatusType.Passing || package.PackageTestResultStatus == PackageAutomatedReviewResultStatusType.Exempted)
+                    && (package.PackageScanStatus == PackageScanStatusType.NotFlagged || package.PackageScanStatus == PackageScanStatusType.Exempted)
+                ))
```
:+1: this has also been addressed.

ferventcoder commented on chocolatey/Website/ViewModels/PackageViewModel.cs:
```diff
@@ -95,16 +99,25 @@ public PackageViewModel(Package package)
+        [Display(Name = "Validator Exempted Reason")]
+        [StringLength(500)]
+        public string ExemptedFromValidatorReason { get; set; }
+
```
👍

ferventcoder commented on chocolatey/Website/Services/PackageService.cs:
```diff
@@ -1133,12 +1137,35 @@ public void ExemptPackageFromTesting(Package package, bool exemptPackage, string
+            {
+                if (!string.IsNullOrWhiteSpace(newComments)) newComments += "{0}".format_with(Environment.NewLine);
+                newComments += "Auto Verification Change - Verfication tests have been set to unknown.";
+            }
```
Good to have this here. Thanks for adding

ferventcoder commented on chocolatey/Website/Services/PackageService.cs:
```diff
@@ -1133,12 +1137,35 @@ public void ExemptPackageFromTesting(Package package, bool exemptPackage, string
             package.PackageValidationResultStatus = PackageAutomatedReviewResultStatusType.Exempted;
             package.PackageValidationResultDate = DateTime.UtcNow;
+            package.ExemptedFromValidatorById = reviewer.Key;
+            package.ExemptedFromValidatorReason = reason;
```
Definitely good to be capturing this for auditing!

ferventcoder commented on chocolatey/Website/Services/PackageService.cs:
```diff
@@ -194,7 +194,7 @@ public PackageRegistration FindPackageRegistrationById(string id, bool useCache)
         public virtual Package FindPackageByIdAndVersion(string id, string version, bool allowPrerelease = true)
         {
-            return FindPackageByIdAndVersion(id, version, allowPrerelease, useCache: true);
+            return FindPackageByIdAndVersion(id, version, allowPrerelease, false);
```
This change seems weird. What specifically was it about and does it get separated to some other issue (I haven't checked commits on where this came in as I'm in mid-review).

gep13 replied:
I made this change locally to ensure that I was always getting the latest package, since I was messing with changes directly in the database.  This wasn't meant to be included in the commit.  Thanks for catching this!

ferventcoder commented on chocolatey/Website/Controllers/PackagesController.cs:
```diff
@@ -191,12 +220,18 @@ public virtual ActionResult DisplayPackage(string id, string version, FormCollec
             {
                 ModelState.AddModelError(String.Empty, "You need to provide comments.");
+
+                TempData["ErrorMessage"] = "There was a problem completing the operation.  Check validation summary for more information.";
```
Oh goodness. We forgot that.... :/

ferventcoder replied:
👍

ferventcoder commented on chocolatey/Website/Controllers/ApiController.cs:
```diff
@@ -568,7 +583,17 @@ public virtual ActionResult SubmitPackageScanResults(string apiKey, string id, s
+
+            packageSvc.UpdateSubmittedStatusAfterAutomatedReviews(package);
+
+            packageSvc.ChangePackageStatus(package, package.Status, package.ReviewComments, message, testReporterUser, testReporterUser, sendMaintainerEmail: true, submittedStatus: package.SubmittedStatus, assignReviewer: false);
```
Little bit concerned about this here as it moved from minor change to ChangePackageStatus which could do something unintended for this section.

gep13 replied:
This is a necessary change, based on bringing the scanner fully into the moderation process.  Both package-validator and package-verifier call this method.

gep13 commented:
@ferventcoder I believe I have addressed all the comments, and have added the last remaining parts to this PR.  Would be great if this could be reviewed again, and ideally taken for a spin.

ferventcoder commented on chocolatey/Website/Infrastructure/Lucene/LuceneSearchService.cs:
```diff
@@ -208,6 +208,7 @@ private static Package PackageFromDoc(Document doc)
                 DownloadCache = doc.Get("DownloadCache"),
                 PackageScanStatusForDatabase = doc.Get("PackageScanStatus"),
                 PackageScanResultDate = packageScanResultDate,
+                PackageScanFlagResultForDatabase = doc.Get("PackageScanFlagResult"),
```
@gep13  PackageIndexEntity sets this I would say.  So you are going to need to set it in there. It's a Lucene Document that is it using to grab that from.

gep13 replied:
Gotcha.  I thought I had caught all the places that this would be required, but looks like I missed this one.

ferventcoder commented on chocolatey/Website/Services/PackageService.cs:
```diff
@@ -1021,31 +1023,33 @@ public void ChangeTrustedStatus(Package package, bool trustedPackage, User user)
                 }
 
-                var trustedPackagePassingAutomatedReview = PackageIsTrusted(package) && passingVerification && passingValidation;
+                var trustedPackagePassingAutomatedReview = PackageIsTrusted(package) && passingScanner && passingVerification && passingValidation && scanFlagResult != PackageScanFlagResultType.Error;
```
👍

ferventcoder commented on chocolatey/Website/Services/PackageService.cs:
```diff
@@ -1078,7 +1082,7 @@ public void ChangePackageTestStatus(Package package, bool success, string result
             packageRepo.CommitChanges();
 
-            var testComments = string.Format("{0} {1}.{2} This is not the only check that is performed so check the package page to ensure a 'Ready' status.{2} Please visit {3} for details.",
+            var testComments = string.Format("{0} {1}.{2} This is not the only check that is performed. Check the package page to ensure the package has reached 'Ready' status.{2} Please visit {3} for details.",
```
I'd almost love for this to report the status of checks that have passed `x/3` so the user knows exactly where they stand. This is where the most confusion for community maintainers is born out of (this message). 

They say it passed automated tests, but in reality it only passed one. So if the message could start out with something like "The package has passed 2/3 automated checks. It just passed x".

gep13 replied:
I have made some changes to this, based on the suggestion that you made for when a package passes validation.  Given that there is a linear path through each stage of automated package testing now, I am not sure if we "need" the `x/3`.  A package either passes each stage, or it gets held in the queue.

gep13 commented on chocolatey/Website/Controllers/ApiController.cs:
```diff
@@ -453,11 +453,11 @@ public virtual ActionResult SubmitPackageValidationResults(string apiKey, string
             {
                 package.PackageValidationResultStatus = PackageAutomatedReviewResultStatusType.Passing;
-                message = "{0} has passed automated validation. It may have or may still fail other checks like testing (verification).".format_with(package.PackageRegistration.Id);
+                message = "{0} has passed automated package validation.".format_with(package.PackageRegistration.Id);
```
:+1:

gep13 commented on chocolatey/Website/Controllers/ApiController.cs:
```diff
@@ -634,7 +649,35 @@ public virtual ActionResult SubmitPackageScanResults(string apiKey, string id, s
+                newSubmittedStatus = PackageSubmittedStatusType.Waiting;
+
+                if (!string.IsNullOrWhiteSpace(message)) message += "{0}".format_with(Environment.NewLine);
+                message += "At least one file within this package had greater than 10 detections.";
```
👍

gep13 commented:
@ferventcoder @vexx32 I believe I have addressed all the comments that have been made as part of the review of this PR.  Please let me know if there are any other comments.  As requested, could either of you confirm whether you have had a chance to take this code for a spin on your environments?  It would be great to get some additional feedback on actually running the code, rather than just a review of the code.

vexx32 reviewed (commented):
@gep13 looks like there's still a couple minor display issues around the **Details** links and sometimes with reasons applied to each stage it goes a bit funny.

![text bunched up with verification exemption](https://user-images.githubusercontent.com/32407840/93123783-a317e180-f696-11ea-83c6-95475211ae8c.png)

Looks like there should be line breaks before and after **Details** to match the others, but it renders the whole thing on a single line.

Also, when there is no exemption comment etc the **Details** link appears in a completely different location, off to the right side. The other **Details** links seem to appear on the line below, we should probably be consistent about that, one way or the other. There seems to be some inconsistencies in their placements...

![two details links in very different locations](https://user-images.githubusercontent.com/32407840/93128901-2d177880-f69e-11ea-80c1-f5750dcf7a40.png)

![all bunched up](https://user-images.githubusercontent.com/32407840/93136332-bd0eef80-f6a9-11ea-9b56-36599fc011f7.png)

I also noticed one case where the `Exempt this version from X` failed to show in the checklist. If both the Validation and Verification checks are skipped/exempted on the same package, it seemed to prevent it from showing a checkbox to exempt from the scanner step. I was still able to advance the states by manually selecting `Exempted` and submitting a response as a moderator though.

Also, not sure of the purpose for the **Package Status** box having `Exempted` states anymore if those are going to be completely separate checkboxes (assuming they always show up when they might be needed, of course).

One last thing -- the first two automatic verification checks state `Passed/Failed/Exempt`, but the scanner uses `Successful/Failed/Exempt(ed?)`. We should probably have them all use the same terminology.

Nothing else of note seems to be breaking, which is lovely! I did hit a few cache issues and think for a bit I might have had some disappearing packages, but they all came back eventually so no issues there either.

gep13 commented:
@designxsteph are you able to provide any assistance in the issues regarding placement of UI that @vexx32 mentions above?

designxsteph commented:
> @designxsteph are you able to provide any assistance in the issues regarding placement of UI that @vexx32 mentions above?

@gep13 @vexx32 I'm finding these display placement issues very strange because I know I have fixed these exact issues before. 🤔 I'll take a look at this today and figure out what's going on!

designxsteph commented:
@gep13 @vexx32 After my commits have been added back in, the display issues seem to be fixed. It would be worth double checking on your local as well to confirm.👍

vexx32 commented:
Nice! Looks good on my end, thanks for that @designxsteph! 💖 

![image](https://user-images.githubusercontent.com/32407840/93389511-40f7e180-f83a-11ea-8220-8da8030d81df.png)
